pageserver/src/tenant.rs

```diff
@@ -1389,13 +1445,107 @@ impl Tenant {
+                        return Err(DeleteTimelineError::from(e))
+                    }
+                    MaybeDone::Pending(rx) => {
+                        // important: upgrading and later resubscription happens while holding the lock
+                        rx.upgrade()
+                    }
+                    MaybeDone::Done(Err(_retryable)) => None,
+                }
+            } else {
+                None
+            };
+
+            if let Some(rx) = maybe_rx {
+                rx
+            } else {
+                // try another time
+                let (tx, rx) = tokio::sync::broadcast::channel(1);
+                let rx = Arc::new(rx);
+                let this = self.clone();
+                let timeline = timeline.clone();
+
+                // TODO: this could be tenant scoped task_mgr task?
+                tokio::spawn({
+                    let rx = rx.clone();
+                    async move {
+                        // to uphold the MaybeDone promise, we keep the channel alive *until* we've
+                        // swapped the MaybeDone::Pending -> MaybeDone::Done value
+                        let tx = tx;
+                        let rx = rx;
+
+                        let res = this.unique_delete_timeline(&timeline).await;
+
+                        {
+                            let mut g = timeline.delete_self.lock().await;
+                            #[cfg(debug_assertions)]
+                            match g.as_ref() {
+                                Some(timeline::MaybeDone::Pending(weak)) => {
+                                    let same = weak
+                                        .upgrade()
+                                        // we don't yet have Receiver::same_channel
+                                        .map(|rx2| Arc::ptr_eq(&rx, &rx2))
+                                        .unwrap_or(false);
+                                    assert!(same, "different channel had been replaced or dropped");
+                                }
+                                other => panic!("unexpected MaybeDone: {other:?}"),
+                            }
+                            *g = Some(timeline::MaybeDone::Done(res.clone()));
+                            drop(rx);
+                        }
+
+                        // now no one can get the Pending(weak) value to upgrade and they only see
+                        // the Done(res).
+
+                        // send the result value to listeners, if any
+                        drop(tx.send(res));
+                    }
+                    .instrument(tracing::info_span!("unique_delete_timeline"))
+                });
+                *g = Some(timeline::MaybeDone::Pending(Arc::downgrade(&rx)));
+                rx
+            }
+            // subscribe while the mutex is held, so we are guaranteed to receive the message,
+            // unless a panic comes.
+            .resubscribe()
```

Comment:
I think those apis are just the same, on receiver it's just called resubscribe. I always found the "sharing a sender to create receivers" weird, but of course the underlying impl allows for that. I think this weirdness is one of the reasons there is now a `resubscribe`.

I don't think this comment is necessary, but if you find I've misunderstood `resubscribe` semantics then perhaps we should change to sharing the sender.

Reply:
>  but if you find I've misunderstood resubscribe semantics then perhaps we should change to sharing the sender.

No, you got them right. It's just the weirdness of the name.

Reply:
It is "resubscribe" because the receiver in `Weak<Receiver>` is by itself locked in to receive all messages due to it being created at the same time as sender at `tokio::sync::broadcast::channel(..)`.

`resubscribe` is not clone where clone would mean "cloned receiver sees messages from same point on as the original." clone is not offered because this would be surprising, I assume.